Aiven Release for v24.8.13.16 LTS

aris-aiven opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- New Feature
- Experimental Feature
- Improvement
- Performance Improvement
- Backward Incompatible Change
- Build/Testing/Packaging Improvement
- Documentation (changelog entry is not required)
- Critical Bug Fix (crash, data loss, RBAC) or LOGICAL_ERROR
- Bug Fix (user-visible misbehavior in an official stable release)
- CI Fix or Improvement (changelog entry is not required)
- Not for changelog (changelog entry is not required)


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

thevar1able committed:
Fix build with clang-19

Backport https://github.com/ClickHouse/ClickHouse/pull/70412

thevar1able committed:
Use libdivide v5.1

nikitamikhaylov committed:
Merge pull request #70691 from ClickHouse/bump-arrow-to-16.1.0

Bump `arrow` to v16.1.0 and `flatbuffers` to v23.5.26

aris-aiven committed:
Fix remote database host:port parsing

This affects MySQL and PostgreSQL when the host is an IPV6.

This commit is applied from the patch file Patch 0002-Fix-remote-database-host-port-parsing.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Advertise host from config for Replicated databases

This commit is applied from the patch file 0006-Advertise_host_from_config_for_Replicated_databases.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Enable internal replication on auto-generated cluster

Internal replication means that when writing to a cluster shard, writes
only go to one replica and data is then copied between replicas using
the normal method.  This patch enables this setting on all Replicated
databases.

This commit was applied from patch file 0007-Enable_internal_replication_on_auto_generated_cluster.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Ignore unreadable sensors

This is a convenience patch to avoid polluting logs on developer laptops
with GPU drivers that advertise non-working thermal sensors.

This commit was applied from the patch file 0008-Ignore_unreadable_sensors.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Fix tcp_port_secure from ZK

This commit was applied from the patch file 0018-Fix_tcp_port_secure_from_zk.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Alter replicated database settings

The Replicated database engine did not support altering its settings.

This was made more necessary with the addition of cluster secrets:
it is useful to be able to rotate secrets quickly without having to
recreate an entire database.

This feature can be used with the following (undocumented) statement:

    ALTER DATABASE `database` MODIFY SETTING cluster_secret='secret'

This commit was applied from the patch file 0020-Alter_Replicated_database_settings.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Added support for protected users

Add an extra flag to each user, called "Protected".

Users with the "Protected" flag can only be created, altered or removed
by users with an extra `PROTECTED_ACCESS_MANAGEMENT` privilege.

This privilege is only meant to be granted to our internal admin user,
and the "Protected" flag will be given to the main service user,
allowing it to exist as an SQL user (which is necessary to have less privileges than
an hardcoded XML users), but not be removable or alterable by the
cluster users.

This commit was applied from the patch file 0021-Protected_users.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Added support for disabling various table engines and table functions

This commit was applied from the patch file 0022-Disable-unsupported-engines.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>
Co-authored-by: Joe Lynch <joe.lynch@aiven.io>
Co-authored-by: Aliaksei Khatskevich <alex.khatskevich@aiven.io>

aris-aiven committed:
Replace MergeTree with ReplicatedMergeTree

This commit was applied from the patch file 0023-Replace-MergeTree-with-ReplicatedMergeTree.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>
Co-authored-by: Joe Lynch <joe.lynch@aiven.io>
Co-authored-by: Dmitry Potepalov <dmitry.potepalov@aiven.io>

aris-aiven committed:
Disallow replication parameters customization

This commit was applied from the patch file 0025-Disallow-replication-parameters-customization.patch

Co-authored-by: Dmitry Potepalov <dmitry.potepalov@aiven.io>

aris-aiven committed:
Disable thread fuzzer

This commit was applied from the patch file 0027-Disable_Thread_Fuzzer.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Added custom SSL settings for PostgreSQL and MySQL

Allow setting the SSL mode and CA path for PostgreSQL and MySQL integrations
when using named collections.

This commit was applied from the patch file 0028-Custom-ssl-settings-postgresql-mysql.patch

Co-authored-by: Joe Lynch <joe.lynch@aiven.io>

aris-aiven committed:
Unlock PostgreSQL database

Fixes an issue where the single mutex in each database
using the PostgreSQL engine is locked while attempting to connect to
PostgreSQL and while running queries.

This is a problem because the same lock is held while reading the
database name and its comment field. This means that a single
failing service integration could make it almost impossible to read
the `system.databases` metadata table, which is used for monitoring
and other critical operations.

This would also make it really hard to drop the problematic integration
database, since the drop operation also does PostgreSQL queries.

The `cached` option of the PostgreSQL database engine does not
improve the situation :
It only caches the table structure but not the table existence, this is
why drop/attach/detach/list operations all need PostgreSQL queries.

This commit was applied from the patch file 0029-Unlock_PostgreSQL_Database.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Added Kafka authentication settings

This patch was applied from the patch file 0031-Kafka_Authentication_Settings.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Restrict "SHOW CREATE database"

Main service users are allowed the SHOW DATABASES access because it is
necessary for their operation. It is implicitly granted by ClickHouse when
giving access to a database.

However, we do not want to give them access to SHOW CREATE DATABASE.
This undocumented but enabled query shows the entire create statement,
unredacted. This is actually a useful feature for superusers, but can leak
credentials to other users.

Also only show the create query in system.tables to users that were able
to create that table.

This commit was applied from the patch files
0032-Restrict_Show_Create_Database.patch and 0033-Restrict_Create_Table_Query.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Tolerate ZooKeeper restart

When ZooKeeper restarts for a reason or another (especially during
version upgrade), it takes around 6 seconds.

ClickHouse fails the currently running query if ZooKeeper is not
available in less than 3 seconds.

This leads to inconsistent database state because ClickHouse does not do
all of its DDL operations atomically. It's really easy to fail a version
upgrade because of that error, and the node requires a manual
intervention to work again.

Hopefully, with more retries and a longer timeout (for a total in
6*2=12sec), ClickHouse should be able to ride the fault and provide a
reliable version upgrade.

This commit was applied from the patch file 0035-Tolerate_ZooKeeper_restart.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Fix ClickHouse restart with replicated table

ClickHouse does not inject the `{shard}` macro value when re-attaching
a table when the server starts.

It used to work in 21.x because we had the custom `{shard_name}`
macro which handled this case.

Now that this patch is removed, we have the `{shard}` bug: a table
can be created both directly and indirectly from the replicated
queries log in ZooKeeper, everything works until the server is
restarted.

When the server restarts, the shard information is not added to
the macro expansion context and the `ATTACH` fails with:

  Caught exception while loading metadata: Code: 62. DB::Exception:
  No macro 'shard' in config while processing substitutions in
  '/clickhouse/tables/{uuid}/{shard}'

Confusingly, `is_replicated_database` is only true for secondary
queries (those coming from the replicated queries log), that why
the extra condition needs to be added to handle the `ATTACH`.

This commit was applied from the patch file 0039-Attach_replicated_table.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Fix ClickHouse settings when adding a replica

ClickHouse global settings can be temporarily modified during a
single session or during a single query.

These settings can alter the behavior of a query, such as allowing
optional or experimental features when creating a table: this is used
for the JSON and Geo data types.

Although they affect table creation, these global settings are
different from the table settings.

When using a Distributed database, a replicated query must use
the same global settings on all servers: the query needs to have
the same effect on all servers.

To be able to do that, ClickHouse adds the global settings to each
entry in the DDL replication log. All servers can read the log and
execute the same query with the same settings.

However, when adding a replica to an existing cluster, the new
replica does not read the replication log. Instead, the replica reads
the table metadata in ZooKeeper. This is done in a function called
`recoverLostReplica` ("lost" means too out-of-sync to apply the
available DDL log entries, which is true for a new replica).

This table metadata contains the table settings but not the global
settings.

Due to that, the new replica will execute all the `CREATE TABLE`
queries with the table settings but without the global settings.

Because of this discrepancy when handling global settings, an
existing replica can create a table using a global experimental
flag, but a new replica will fail to create the same table.

Ideally, we'd apply all global settings when recovering a lost
replica, but that information is not available in the table metadata.

Instead, we can detect the queries executed by `recoverLostReplica`
and selectively disable the guardrails preventing the creation of
the table (these guardrails are pointless anyway since the table
already exists in all other replicas).

Since ClickHouse 23.8, most settings are correctly enabled, so this
patch became much smaller, but at least one setting is still missing.

Similarly, If the table was initially created with flatten_nested=0,
this caused a failure when adding a new replica to an existing
cluster: the table would be created with flatten_nested=1 on the new
replica and the schemas would not match.

This commit was applied from the patch file 0040-Recover_replica_settings.patch

aris-aiven committed:
Disable replicas_status endpoint

This commit was applied from the patch file 0042-Disable_replicas_status.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Fix swap drift

Even with low swappiness, sometimes ClickHouse pages go to swap.

By itself, it's not a problem, unused pages can move to swap without
any issue. However, this interacts badly with the memory tracker of
ClickHouse.

This memory tracker is in charge of limiting the total memory usage
of the server without interrupting the service (unlike the OOM killer
or an earlyOOM daemon). Instead, individual queries are delayed
or rejected until more memory is available.

The tracker normally only looks at the RSS of the process, which
excludes the swap.

If part of ClickHouse memory is moved to swap, then the tracker stops
counting that memory towards the limit and allows more memory
allocations. Then a little bit more memory is moved to swap, and this
cycle keeps going until all swap is consumed.

In case of low swappiness configuration of the system it can take days
or weeks before ClickHouse consumes all swap.

At that point, ClickHouse is still below its configured maximum ram
usage, as configured. But, if some swapped page suddenly needs to
move back to ram, then we have a real problem: there is not enough
room in ram because all processes were allocated a portion of it
and there is no room in swap because ClickHouse slowly consumed
everything. We have to move things around but no room to move,
it's swap hell.

Fix it by taking in account the swapped memory of the process in
addition to its RSS.

To get this data, ClickHouse needs to be modified to look at
`/proc/self/status` instead of `/proc/self/statm`, and adapt to the
different file format. (statm does not contain swap information).

As a bonus, the swapped memory is exposed as a metric like the other
memory metrics already present.

This commit was applied from the patch file 0044-Fix_swap_drift.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Allow custom CA in S3

For testing purposes, we want to be able to connect to an object storage
server that does not have a certificate signed by a public authority,
while still testing it with HTTPS.

ClickHouse does not have that option, but the underlying library
(POCO), supports it, just pass the parameter around, from the disk
settings to the library.

This commit was applied from the patch file 0047-Custom_S3_certificate_authority.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Fix IPv6 Azure object storage host

The Azure SDK does not fully support IPv6 in hostnames, mainly
because the escaping brackets are not parsed and removed at
the right time.

The second fix is to detect IPv6 in hosts and validate them by
parsing them instead of rejecting them because they don't
look like DNS name segments.

This commit was applied from the patch file 0048-Fix_IPv6_Azure_object_storage_host.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Allow custom CA in Azure

For testing purposes, we want to be able to connect to an object
storage server that does not have a certificate signed by a public
authority, while still testing it with HTTPS.

ClickHouse does not have that option, but the underlying library
(Curl, below Azure), supports it, just pass the parameter around,
from the disk settings to the library.

This commit was applied from the patch file 0049-Custom_Azure_certificate_authority.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Fix unbounded replication queue

If a replica is struggling to execute all tasks in its replication
queue, the queue can grow to a huge size and is even slower to manage
when its larger (it's an `std::list`).

This queue is also present in ZooKeeper, for each replica.

There is also a shared queue, which is called the log. It's used to
communicate changes between replicas. One replica will push an item on
the log as part of executing a query, then all other replicas will read
that log item and repush it in their own queue.

(This is an approximation, the queue items are not identical to the log
items, but filling the log fills the queues).

Each replica monitors the log size and which replicas are deemed "lost"
when they are too late to consume log items. However, the replicas do
not monitor the size of the queue of the other replicas.

If a replica is pushing lots of items in the log and another replica is
able to copy them to their own queue, but too slow to consume the queue
items, then we have a problem.

At some point, if the imbalance persists, ZooKeeper will crash because
it won't have enough memory to hold all the queues. It will be very hard
to recover: ZooKeeper itself often corrupt its state when OOMing.
Manually cleaning hundred of thousand of items in the queues to know
which ones can be removed and which one are necessary to keep the
replicated data in a consistent state is almost impossible.

To fix that, we reuse the infrastructure that delays insert queries when
there are too many parts. The same logic is applied when any replica queue
size is too large for the table, or the grand total of all replicas' queue
sizes, over all tables, is too large. We also do the same to fail queries
if either counters reach a second threshold, exactly like it's done with
the parts count.

In each replica, a thread is updating a local copy of the maximum queue
size in all replicas. A map is added to the context to keep track of these
per-storage maxima. These values are then used to gate the inserts. We do
it like that to avoid adding many ZooKeeper queries in the hot path, and we
don't need a very accurate queue size to get the feedback loop we need.

This commit was applied from the patch file 0050-Fix_unbounded_replication_queue.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Add support for Azure object storage path prefix

Azure storage does not support having a prefix before the stored
object key.

This makes it unusable for us when we use shared buckets with
prefixes for projects and backup sites.

This commit was applied from the patch file 0051-Azure_object_storage_prefix.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Allow decreasing number of Kafka consumers to zero

Decrease the lower bound on the no. kafka consumers.

This allows settings it to zero, effectively disabling kafka consumers globally without
relying on SQL statement execution, which would allow users to interfere adversely with
management operations.

This commit was applied from the patch file 0053_Decrease_kafka_num_consumers_bound.patch

aris-aiven committed:
Fix metadata with encrypted disk

Add functions missing in DiskEncrypted implementation.

This class has a long record of our patches. We found many missing
functions, and some of them are fixed in the upstream already.

Added functions:
- `getRemotePathsRecursive`: prevents using some of the system tables that were
  relying on this function to expose debugging information. Will
  be deleted in 24.8.
- `getRefCount`: absence of this method leads to preliminary ZeroCopy znode
  deletion during unfreeze and hanging VM recycling.

This commit was applied from the patch file 0055-Encrypted_disk_metadata.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Fix default profile escape

We want to restrict settings using the default profile, while having
less restricted settings for our superadmin user with the admin
profile.

We also grant the user the capability to create its own users, roles
and settings profile.

Normally, you cannot switch to settings profiles that are less
restrictive than your current profile, ClickHouse verifies the new
profiles against the constraints of the current profile.

However, when you can create new users, roles, etc. You can create
users that use the admin profile, then directly connect with this
user.

This way you can skip the verification that happens while switching
profiles and escape the constraints of the default profile.

Prevent that by enforcing that all created profiles must inherit from
the default profile or one of its descendants, unless the current
has a settings allowing to skip that: `allow_non_default_profile`.

This works for us because the admin profile is pre-created using
users.xml and not via SQL, so it can already exist and be configured
with `allow_non_default_profile` set to true, while the pre-created
default profile has `allow_non_default_profile` set to false.

This commit was applied from the patch file 0056-Unescapable_default_profile.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Check MergeTree settings constraints before enqueuing DDL queries

The constraints were previously only checked when executing the query
from the replicated DDL queue. At that point the DDLWorker was using
the system profile for settings constraints.

This meant that a non-admin user, using the default profile, could use
a MergeTree settings that was not allowed by its profile.

Fix that by checking the MergeTree settings constraints before the
query is enqueued, when we still have the query context attached to
the user running the query.

We check both Replicated*MergeTree and apparently non-replicated
MergeTree because this check happens before we rewrite the query
to enforce Replicated*MergeTree engine types.

This commit was applied from the patch file 0057-Early_MergeTree_settings_constraints.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Add Backup disk type

This adds a new disk type that wraps an object storage disk and
enables coordination with our backup system.

When this disk type is used as a wrapper on top of an object storage
disk, the files in object storage are not really deleted.

Instead, a local file is created for each file in object storage that
should be deleted.

Our backup system will then be in charge to do garbage collection
and only delete files are are not referenced by any backup.

The file names of the deletion markers are like this:
CLICKHOUSE_DIR/disks/remote_backup/backup/ESCAPED_OBJECT_PATH

The object path is escaped to simplify directory management and uses
the same escape mechanism that ClickHouse uses in many places to
reference files (essentially escape encoding).

In case Backup disk is wrapping a Cache disk then it lets the remove
calls pass

This commit was applied from the patch file 0059-Backup_disk.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Support per-table schema registry with authentication

Add an extra option for tables with the Kafka engine to pass the URL
of a schema registry.

Also add support for schema registry basic authentication.

This commit was applied from the patch file 0061-Avro_schema_registry.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Fix ClickHouse trying to read non-existent certificate files

If both the certificate file and key file are defined, even when they are
not used, ClickHouse will try to read and parse them. We can't even
work around the issue by creating empty files.

This commit was applied from the patch file 0064-Lazy_certificates.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Fix uncaught exception if S3 storage fails

If an exception happened while copying data, then the finalize call
was not done explicitly.

The finalize was then be done while the out object is destructed and
that led to a fatal error, killing the entire server.

To fix that, catch the exception that might happen while copyData is
running, run finalize there and rethrow.

The rethrow is very important, the finalize after a failing copyData
can also not-throw, in which case the initial error would be
forgotten and ClickHouse would believe the copy succeeded.

This commit was applied from the patch file 0068-Uncaught_S3_Exception.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Enforce SSL in the MySQL handler

This commit was applied from the patch file 0071-MySQL_enforce_SSL.patch

Co-authored-by: Joe Lynch <joe.lynch@aiven.io>

aris-aiven committed:
Add support for integration metadata to named collections validation

We need to keep track of existing named collections. In order to do
that, we wish to add some metadata to each collection. The metadata keys
is added as optional collection parameters for the storages and
functions that validate the keys.

This commit was applied from the patch file 0074-Named_Collections_Integration_Metadata.patch

Co-authored-by: Aris Tritas <aris.tritas@aiven.io>

aris-aiven committed:
Replicate more ALTER TABLE MOVE PARTITION queries.

We require these queries to be replicated to all nodes of the database.

This commit was applied from the patch file 0075-Replicate-move-partition.patch

Co-authored-by: Joe Lynch <joe.lynch@aiven.io>

aris-aiven committed:
Add early fetch pool

When adding a replica to an existing cluster, the replica will add
many GET_PART tasks to its replication queue. These tasks are in
charge of downloading the data that existed before the creation of
the replica.

Meanwhile, if the cluster is still receiving updates, new GET_PART
tasks will be created to add the extra data.

These new parts are usually small, and in a cluster where the inserts
are a bit too frequent, we can have a lot of them.

However, there is a limit to the number of GET_PART tasks that a
replica will simultaneously execute (by default 8).

The early GET_PART tasks, for existing data, are often very large and
can take a lot of time to execute.

If there are more existing parts than the maximum number of
simultanenous downloads, then the early parts will take all slots in
the pool for download tasks and the small additional GET_PART tasks
will have to wait for the very large early parts to be downloaded.

The accumulation of many small GET_PART tasks then becomes an issue
for ZooKeeper, each of these tasks is a znode and it's quite easy to
end up with >100k pending tasks for a single table.

We cannot simply increase the download pool size. A large cluster
will always have more existing large parts than the pool size, unless
we configure the pool with an unreasonably large size.

However, we can split the pool in two parts: one pool for the
existing parts that are added early to sync the replica, and another
pool for the normal GET_PART tasks that happen because of normal
inserts.

This commit was applied from the patch file 0077-Early_fetches_pool.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Include metadata_version.txt when freezing

This file was added to each part folder to help with concurrency issues
when mutating the table schema, but it was not added to the frozen
files.

We need to include this file in the frozen shadow/ folder because we
will use these file to know what to backup.

When using object storage, this file is actually a pointer to the file
in object storage. We use these pointers to know which files are still
referenced by ClickHouse and should not be deleted.

By omitting this file, we would not know about it, and delete the
metadata_version.txt file in object storage. This was causing latent
issues, ClickHouse doesn't immediately notice the missing file, but
instead complains loudly when adding a new replica: the new replica
tries to download this file when syncing from existing replicas.

This commit was applied from the patch file 0084-Freeze_metadata_version.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Fix alter order by

`alter code` is detached from `create table` code, which makes it
necessary to copy field initializaiton logic. This commit makes
`alter table` to produce the same `sorting key` ZooKeeper metadata as
`create table`.

This commit was applied from the patch file 0087-Fix-alter-order-by.patch

Co-authored-by: Aliaksei Khatskevich <alex.khatskevich@aiven.io>

aris-aiven committed:
Changes for ClickHouse dictionary source, both remote and local.

Remote:
* Allow use of addresses_expr in named collection configuration.
* Enforce `secure` parameter for ClickHouse dictionary source when not
  using named collections.

Local:
* New server setting `dictionary_user` to specify which user
  to run the dictionary query, rather than just the default user
  (since we don't have a default user)

This commit was applied from the patch file 0089-Dictionary-source-ClickHouse.patch

Co-authored-by: Joe Lynch <joe.lynch@aiven.io>

aris-aiven committed:
Enforce HTTPS for URL storage and HTTPDictionarySource

This commit was applied from the patch file 0091-HTTP-dictionary-source.patch

Co-authored-by: Joe Lynch <joe.lynch@aiven.io>

aris-aiven committed:
Remove all references to ClickHouse Cloud

This commit was applied from the patch file 0094-strip-cloud-references.patch

Co-authored-by: Joe Lynch <joe.lynch@aiven.io>

aris-aiven committed:
Handle ZooKeeper error while creating table

Catch additional ZK connection errors while creating table
and handle the case of dropping empty table directories

This commit was applied from the patch file 0095-Handle-zk-error-while-creating-table.patch

Co-authored-by: Sumit Jamgade <sumit.jamgade@aiven.io>

aris-aiven committed:
Fix ZK node leak after create delete table

Minimal backport of https://github.com/ClickHouse/ClickHouse/pull/69574.

This commit was applied from the patch file 0098-create-delete-znode-leak.patch

Co-authored-by: Joe Lynch <joe.lynch@aiven.io>

aris-aiven committed:
Zero copy fixes

1. Fix sharding race condition with zero copy.  Locks are shared between parts
  with the same name on different shards.  This can lead to a race condition
  when committing the same part on different shards for the znode
  `/clickhouse/zero_copy/zero_copy_s3/6c758aa1-6d33-4801-bd9c-740e0cea7307/all_0_0_0`.
  The fix is to create this znode before committing the part and using createIfNotExists
  style operations.
  Especially prevelent on SYSTEM RESTORE REPLICA on the first partition.
2. Remove zero copy replication error when dropping table - occurs Astacus has left
  files in the table directory after a failed restore.

This commit was applied from the patch file 0099-zero-copy-backup-restore-fixes.patch

Co-authored-by: Joe Lynch <joe.lynch@aiven.io>

aris-aiven committed:
Add kafka_auto_reset_mode and kafka_date_time_input_format settings

kafka_auto_reset_mode was only configurable by topic through the global
config file and kafka_date_time_input_format was only configurable through
the global config file or the context (but context settings affecting
CREATE query do not work well with replication/restoration).

This commit was applied from the patch file 0062-Kafka_extra_settings.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Allow delegating S3 signature to a separate process

The process is accessed over HTTP and the URL to access it can be
configured with the signature_delegation_url parameter in the S3 disk
configuration.

ClickHouse will make a POST request with a json body:
{
    "canonicalRequest": "PUT\n...",
}
And expects a JSON response with status 200 and this structure:
{
    "signature": 01234567890abcdefg..."
}

The canonical request matches the format defined by AWS for
signatures and all the extra parameters allow computing the full
string to sign.

The canonical request also contains all the required  information
(path, host, operation...) to let the proxy decide of the request is
allowed and can be signed. It's also enough information to know
the class and cost of the request.

This commit was applied from the patch file 0066-Delegated_signature.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Allow delegating Azure signature to a separate process

The process is accessed over HTTP and the URL to access it can be
configured with the signature_delegation_url parameter in the Azure
disk configuration.

ClickHouse will make a POST request with a json body:
{
    "stringToSing": "..."
}
And expects a JSON response with status 200 and this structure:
{
    "signature": 01234567890abcdefg..."
}

The string to sign matches the format defined by Azure for
signatures.

The canonical request also contains all the required  information
(path, operation...) to let the proxy decide of the request is
allowed and can be signed. It's also enough information to know
the cost of the request.

This commit was applied from the patch file 0067-Delegated_signature_azure.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Add per-server override for max bytes to merge/mutate

During a maintenance upgrade, we do not wait for the completion of
merges and mutations since they are not required to make sure we have
all the data from the previous nodes.

We do wait for `GET_PART` and similar tasks, since we need the new
nodes to get parts from the old nodes. This is implemented using
`SYSTEM SYNC REPLICA ... LIGHTWEIGHT`.

Some merges and mutations prevent the execution of `GET_PART` tasks
that overlap the range of the merge or the mutation. So, even if we
are not waiting for `MERGE_PARTS` tasks, these tasks can slow down
the completion of a maintenance upgrade.

We need to execute `MERGE_PARTS` tasks to avoid having too many
parts in the same partition or table. We also need to execute the TTL
delete rules and ensure disk usage does not grow too much (TTL
deletes are implemented as a subtype of the merge tasks).

A possible tradeoff is to only execute merge tasks if they are not too large,
the small merges are the ones that keep the number of parts low when
the are many small INSERT queries.

We can't use the normal merge tree settings like
`max_bytes_to_merge_at_min/max_space_in_pool` because they would be
applied with `ALTER TABLE` which would step onto user-managed objects
and become persisted as part of the table definition. They are also
replicated, we can't set them only on new nodes.

The patch implements per-server overrides, that we can use to limit
the size of merge and mutate tasks, both when a node creates new
tasks and when it decides which task to execute.

This is only usable with a `LIGHTWEIGHT` sync: if we don't execute
some tasks, we also need to not wait for them, or we would wait
forever. `SYNC REPLICA` with the `LIGHTWEIGHT` flag allows us to
not wait for these tasks.

This commit was applied from the patch file 0078-Global_merge_and_mutate_override.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Fixes when named collections are deleted or modified

Currently, when a named collection is modified, the table engine keeps
using the previous credentials until the CH is restarted.  When a named
collection is deleted, the table is still accessible until a server
restart, for which CH crashes.

This commit fixes two issues with named collections for the following
storage types:
* PostgreSQL
* MySQL

When a table uses a storage engine with a named collection in the
definition eg PostgreSQL(`psql_named_collection`) we require the
following:
1. If `psql_named_collection` is deleted, the table should still
  exist but be unusable; an error is thrown on read or write.
  The database should also be able to restart in this case!
2. If `psql_named_collection` is modified, the table should be
  reloaded with the new configuration.

This commit was applied from the patch file 0081-named-collections-reloading.patch

Co-authored-by: Joe Lynch <joe.lynch@aiven.io>

aris-aiven committed:
Add read-only setting to KeeperMap storage

This commit was applied from the patch file 0100-Add-read-only-setting-to-KeeperMap-storage.patch

Co-authored-by: Salvatore Mesoraca <salvatore.mesoraca@aiven.io>

aris-aiven committed:
Allow disabling of individual dictionary sources

This commit was applied from the patch file 0086-Disable-dictionary-sources.patch

Co-authored-by: Joe Lynch <joe.lynch@aiven.io>

aris-aiven committed:
Added support for named collection reloading for S3 and Data Lake table engines

This commit was applied from the patch file 0083-s3-named-collections-reloading.patch

Co-authored-by: Joe Lynch <joe.lynch@aiven.io>

aris-aiven committed:
Named Collections reloading for ObjectStorageQueue

aris-aiven committed:
Fix native macOS build (backport https://github.com/ClickHouse/ClickHouse/pull/70671)

aris-aiven committed:
Remove unused table function registration declarations

Khatskevich committed:
Allow avnadmin creating database using sql

There are 2 main changes:
1. Allow `avnadmin` calling `CREATE DATABASE`, which overrides
during execution phase, ensuring correct parameters are passed and
running it with `aiven` privileges.
2. Introduce `GRANT DEFAULT REPLICATED DATABASE PRIVILEGES` statement,
which will give a user a default set of privileges for a database.
This is necessary to maintain a common source of truth while creating
databases from different environments.

aris-aiven committed:
Multiple changes in PostgreSQL dictionary

* Allow using DDL created named collection for PostgreSQL dictionary.
  Currently this fails upstream.
* Ensure that TLS works.
* Enforce using named collection.

This commit was applied from the patch file 0090-PostgreSQL-dictionary.patch

Co-authored-by: Joe Lynch <joe.lynch@aiven.io>

aris-aiven committed:
Skip attempt to create a container in azure blob storage

This commit was applied from patch 0104-skip-attempt-to-create-a-container-in-azure-blob-storage

Co-authored-by: Tilman Moeller <tilman.moeller@aiven.io>

aris-aiven committed:
Fix IPv6 S3 object storage host

The AWS SDK does not fully support IPv6 in hostnames, mainly
because the escaping brackets are not parsed and removed at
the right time.

This commit was applied from the patch file 0046-Fix_IPv6_S3_object_storage_host.patch

Co-authored-by: Kevin Michel <kevin.michel@aiven.io>

aris-aiven committed:
Fix clustered Azure Blob Storage function query parser

Failing to recognize the format leads to args being pushed on the vector,
 and ultimately a faulty Azure Blob client being created.

aris-aiven committed:
Add support for disabling TimeSeries table engine and function

aris-aiven committed:
Fix AzureObjectStorage constructor call with correct container client

We weren't passing down the container client we just created.
The calls should be more or less equivalent.

joelynch commented on src/Databases/DatabaseReplicated.cpp:
```diff
@@ -62,6 +62,7 @@ namespace ErrorCodes
     extern const int NO_ACTIVE_REPLICAS;
     extern const int CANNOT_GET_REPLICATED_DATABASE_SNAPSHOT;
     extern const int CANNOT_RESTORE_TABLE;
+    extern const int QUERY_NOT_ALLOWED;
```
This appears unused in this commit

joelynch commented on src/Parsers/Access/ParserCreateUserQuery.cpp:
```diff
@@ -501,6 +516,13 @@ bool ParserCreateUserQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expec
+        bool new_protected_entity;
+        if (!protected_entity && parseProtected(pos, expected, new_protected_entity))
+        {
+            protected_entity = std::move(new_protected_entity);
```
nit: std::move here doesn't do anything

joelynch commented on src/Parsers/Access/ParserCreateUserQuery.cpp:
```diff
@@ -568,6 +590,7 @@ bool ParserCreateUserQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expec
     query->hosts = std::move(hosts);
     query->add_hosts = std::move(add_hosts);
     query->remove_hosts = std::move(remove_hosts);
+    query->protected_entity = std::move(protected_entity);
```
same here

joelynch commented on src/Interpreters/InterpreterCreateQuery.cpp:
```diff
@@ -1974,6 +1975,20 @@ BlockIO InterpreterCreateQuery::execute()
+                return {};
+            }
+        }
+    }
```
Is this still needed?

aris-aiven replied:
Good catch, removed

joelynch commented on src/Storages/StorageDictionary.cpp:
```diff
@@ -170,18 +169,7 @@ Pipe StorageDictionary::read(
-    if (!has_dict_get && !has_select)
-        local_context->checkAccess(AccessType::dictGet, dictionary->getDatabaseOrNoDatabaseTag(), dictionary->getDictionaryID().getTableName());
-
+    local_context->checkAccess(AccessType::dictGet, dictionary->getDatabaseOrNoDatabaseTag(), dictionary->getDictionaryID().getTableName());
```
No need to revert here IMO

Commit:
base: disable custom musl/getauxval.c for aarch64

This commit partially backports https://github.com/ClickHouse/ClickHouse/pull/70778
thus fixing the build failure on aarch64 when using sanitizers:
https://github.com/ClickHouse/ClickHouse/issues/59332

DDB-1707

joelynch reviewed (approved):
All comments addressed 👍
